### ISSUE TITLE
Release v0.4.22

### ISSUE THREAD

**Issue by Stebalien** [ipfs/go-ipfs#6506]
# go-ipfs 0.4.22 release

We're releasing a PATCH release of go-ipfs based on 0.4.21 containing some critical fixes.

The past several releases have been shaky and the network has scaled to the point where where small changes can have a wide-reaching impact on the entire network. To keep this situation from escalating, we've put a hold on releasing new features until we can improve our [release process](https://github.com/ipfs/go-ipfs/pull/6482) (which we will be trialing in this release) and [testing procedures](https://github.com/ipfs/go-ipfs/issues/6483).

**Current RC**: [v0.4.22-rc1](https://dist.ipfs.io/go-ipfs/v0.4.22-rc1)
**Install**: `ipfs update install v0.4.22-rc1`

## 🗺 What's left for release

* [ ] Get the interop tests passing (https://github.com/ipfs/interop/issues/71)

## 🔦 Changelog

This release includes fixes for the following regressions:

1. A major bitswap throughput regression introduced in 0.4.21 ([ipfs/go-ipfs#6442](https://github.com/ipfs/go-ipfs/issues/6442)).
2. High bitswap CPU usage when connected to many (e.g., 10,000) peers. See [ipfs/go-bitswap#154](https://github.com/ipfs/go-bitswap/issues/154).
2. The local network discovery service sometimes initializing before the networking module, causing it to announce the wrong addresses and sometimes complain about not being able to determine the IP address) ([ipfs/go-ipfs#6415](https://github.com/ipfs/go-ipfs/pull/6415)).
   
It also includes fixes for:

1. Pins not being persisted after `ipfs block add --pin` ([ipfs/go-ipfs#6441](https://github.com/ipfs/go-ipfs/pull/6441)).
2. Concurrent map access on GC due to the pinner ([ipfs/go-ipfs#6419](https://github.com/ipfs/go-ipfs/pull/6419)).
3. Potential pin-set corruption given a concurrent `ipfs repo gc` and `ipfs pin rm` ([ipfs/go-ipfs#6444](https://github.com/ipfs/go-ipfs/pull/6444)).
4. Build failure due to a deleted git tag in one of our dependencies ([ipfs/go-ds-badger#64](https://github.com/ipfs/go-ds-badger/pull/65)).

## ✅ Release Checklist

For each RC published in each stage:

- version string in `version.go` has been updated
- tag commit with vX.Y.Z-rcN
- upload to dist.ipfs.io
  1. Build: https://github.com/ipfs/distributions#usage.
  2. Pin the resulting release.
  3. Make a PR against ipfs/distributions with the updated versions, including the new hash in the PR comment.
  4. Ask the infra team to update the DNSLink record for dist.ipfs.io to point to the new distribution.

Checklist:

- [x] **Stage 1 - Internal Testing**
  - [x] Feature freeze. If any "non-trivial" changes (see the footnotes of [docs/releases.md](https://github.com/ipfs/go-ipfs/tree/master/docs/releases.md) for a definition) get added to the release, uncheck all the checkboxes and return to this stage.
  - [x] CHANGELOG.md has been updated
    - use [`./bin/mkreleaselog`](https://github.com/ipfs/go-ipfs/tree/master/bin/mkreleaselog) to generate a nice starter list
  - [x] Automated Testing (already tested in CI) - Ensure that all tests are passing, this includes:
    - [x] unit, sharness, cross-build, etc (`make test`)
    - [x] lint (`make test_go_lint`)
    - [x] [interop](https://github.com/ipfs/interop#test-with-a-non-yet-released-version-of-go-ipfs)
    - [x] [go-ipfs-api](https://github.com/ipfs/go-ipfs-api)
    - [x] [go-ipfs-http-client](https://github.com/ipfs/go-ipfs-http-client)
  - [-] Network Testing:
    - [-] test lab things - Not Ready.
  - [x] Infrastructure Testing:
    - [x] Deploy new version to a subset of Bootstrappers
    - [-] Deploy new version to a subset of Gateways -- skipped as these are currently running a special fork.
    - [x] Deploy new version to a subset of Preload nodes
    - [x] Collect metrics every day. Work with the Infrastructure team to learn of any hiccup
  - [x] IPFS Application Testing -  Run the tests of the following applications:
    - [x] WebUI 
    - [x] IPFS Desktop - @hacdias
    - [x] IPFS Companion - @lidel
    - [x] NPM on IPFS - @achingbrain
- [x] **Stage 2 - Public Beta**
  - [x] Reach out to the IPFS _early testers_ listed in [docs/EARLY_TESTERS.md](https://github.com/ipfs/go-ipfs/tree/master/docs/EARLY_TESTERS.md) for testing this release (check when no more problems have been reported). If you'd like to be added to this list, please file a PR.
  - [x] Reach out to on IRC for beta testers.
  - [x] Run tests available in the following repos with the latest beta (check when all tests pass):
    - [x] [orbit-db](https://github.com/orbitdb/orbit-db)
      - two tests failed on both js-ipfs and go-ipfs.
- [x] **Stage 3 - Soft Release**
  - [x] Documentation
    - [x] Ensure that [CHANGELOG.md](https://github.com/ipfs/go-ipfs/tree/master/CHANGELOG.md) is up to date
    - [x] Ensure that [README.md](https://github.com/ipfs/go-ipfs/tree/master/README.md)  is up to date
    - [-] ~Ensure that all the examples we have produced for go-ipfs run without problems~ *infeasible without automation* https://github.com/ipfs/docs/issues/232
    - [x] Update HTTP-API Documentation on the Website using https://github.com/ipfs/http-api-docs
  - [x] Invite the IPFS [_early testers_](https://github.com/ipfs/go-ipfs/tree/master/docs/EARLY_TESTERS.md) to deploy the release to part of their production infrastructure.
  - [ ] Invite the wider community through (link to the release issue):
    - [ ] [discuss.ipfs.io](https://discuss.ipfs.io/c/announcements)
    - [ ] Twitter
    - [ ] IRC
- [x] **Stage 4 - Release**
  - [x] Final preparation
    - [x] Verify that version string in [`repo/version.go`](https://github.com/ipfs/go-ipfs/tree/master/repo/version.go) has been updated
    - [x] tag commit with vX.Y.Z
    - [x] update release branch to point to release commit (`git merge vX.Y.Z`).
    - [x] Release published
      - [x] to [dist.ipfs.io](https://dist.ipfs.io)
      - [x] to [npm-go-ipfs](https://github.com/ipfs/npm-go-ipfs)
      - [x] to [chocolatey](https://chocolatey.org/packages/ipfs)
      - [x] to [github](https://github.com/ipfs/go-ipfs/releases)
  - [x] Publish a Release Blog post (at minimum, a c&p of this release issue with all the highlights, API changes, link to changelog and thank yous)
  - [x] Broadcasting (link to blog post)
    - [ ] Twitter
    - [x] IRC
    - [x] [Reddit](https://reddit.com/r/ipfs)
    - [x] [discuss.ipfs.io](https://discuss.ipfs.io/c/announcements)
    - [x] Announce it on the [IPFS Users Mailing List](https://groups.google.com/forum/#!forum/ipfs-users)

## ❤️ Contributors

* @hannahhoward for fixing both bitswap issues.
* @sanderpick for catching and fixing the local discovery bug.
* @campoy for fixing the build issue.

Would you like to contribute to the IPFS project and don't know how? Well, there are a few places you can get started:

- Check the issues with the `help wanted` label in the [go-ipfs repo](https://github.com/ipfs/go-ipfs/issues?q=is%3Aopen+is%3Aissue+label%3A%22help+wanted%22)
- Join an IPFS All Hands, introduce yourself and let us know where you would like to contribute - https://github.com/ipfs/team-mgmt/#weekly-ipfs-all-hands
- Hack with IPFS and show us what you made! The All Hands call is also the perfect venue for demos, join in and show us what you built
- Join the discussion at [discuss.ipfs.io](https://discuss.ipfs.io/) and help users finding their answers.
- Join the [Go Core Dev Team Weekly Sync](https://github.com/ipfs/team-mgmt/issues/650) and be part of the Sprint action!

## ⁉️ Do you have questions?

The best place to ask your questions about IPFS, how it works and what you can do with it is at [discuss.ipfs.io](http://discuss.ipfs.io). We are also available at the `#ipfs` channel on Freenode, which is also [accessible through our Matrix bridge](https://riot.im/app/#/room/#freenode_#ipfs:matrix.org).

---

This release is currently being readied in #6484.

**Comment by campoy** [comment]
Getting some love when I was actually the cause of the breakage?
You're a classy project 😄 

❤️

**Comment by Stebalien** [comment]
@campoy I've done the same thing.

**Comment by Stebalien** [comment]
Stage 1... done!
Stage 2...

Early testers:

1. Please read https://github.com/ipfs/go-ipfs/blob/master/docs/EARLY_TESTERS.md (which now describes the expectations) and confirm that you're willing to participate (@obo20, I assumed you wanted to be on this list).

2. go-ipfs v0.4.22-rc1 has passed all internal testing and is ready for public beta testing. Please try it out on your test infra (if any) and run your tests suites/apps against it.

3. Please confirm when you've done all relevant testing so we can move on to stage 3.

- [ ] Infura (@MichaelMure)
- [ ] Textile (@sanderpick)
- [ ] Pinata (@obo20)
- [ ] RTrade (@postables)
- [ ] QRI (@b5)
- [ ] Siderus (@koalalorenzo)

This release adds no new features, just some critical fixes applied to v0.4.21. See the highlights in the issue description for a list of changes.

**Comment by sanderpick** [comment]
Call me dangerous, but we've been running ahead of the releases because of a pending cluster integration, which depends on [go-libp2p-core](https://github.com/libp2p/go-libp2p-core/). I'm not sure it makes sense to backpedal and test with v0.4.22-rc1, but I can give that a shot this weekend.

**Comment by Stebalien** [comment]
@sanderpick don't bother. All the changes in 0.4.22-rc1 are _also_ in master. I'll count that as a sign-off.

**Comment by obo20** [comment]
Things are working well on our end @Stebalien. Nothing major to report.

**Comment by koalalorenzo** [comment]
Thx @Stebalien, have some suggestion from somebody with both feet still on planet earth 🙃

1. Use [Semantic Versioning](https://semver.org) to understand easily how changes are impacting our systems
2. Include in the ChangeLog beta/rc versions so we know what to test ( see [Orion's](https://gitlab.com/siderus/orion/shell/blob/master/CHANGELOG.md) and [Keep a Changelog](https://keepachangelog.com/en/1.0.0/))
3. Is it possible to include the builds in GitHub when tagging rc or beta releases? (there is a feature to mark it as a "pre-release". Reason: our pipelines are not downloading binaries from dist.ipfs.io because it is always failing (timeout mostly, due to being backed by IPFS), while GitHub is more reliable. 🤷‍♂️ 

_I still don't know why Protocol Labs has its own non-conventional things_ 😅 and here is a cute cat:

![cute-small-cat-wallpaper](https://user-images.githubusercontent.com/33528/61594556-7c266280-abed-11e9-8197-badb0c294647.jpg)

**Comment by postables** [comment]
Looks good in CI tests. No new or negative issues spotted in our development environment so far. 

How long do we have to give our final analysis? Ideally I'd like to test things for a week or so in dev but if that's too long thats understandable

**Comment by Stebalien** [comment]
> Use Semantic Versioning to understand easily how changes are impacting our systems.

The next release with features will be v0.5.0 so we can clearly distinguish between patch releases and feature releases. For some context:

1. 0.5.0 was _supposed_ to be the official "beta" release of IPFS. Hence the whole "stuck on 0.4.x" thing.
2. Historically, we've used minor releases to indicate major breaking changes (0.3.0 -> 0.4.0 broke network compatibility). This is actually pretty common in pre-1.0 software to clearly indicate breaking changes.

> Include in the ChangeLog (the file) beta/rc versions so we know what to test ( see Orion's and Keep a Changelog)

There's a changelog in the release PR. The previously named "highlights" now named "changelog" section in the issue body is a complete changelog (sorry for the confusion).

> Is it possible to include the builds in GitHub when tagging rc or beta releases? (there is a feature to mark it as a "pre-release").

Sure.

> Reason: our pipelines are not downloading binaries from dist.ipfs.io because it is always failing (timeout mostly, due to being backed by IPFS), while GitHub is more reliable. 

Is this still happening (i.e., since July)? 

> I still don't know why Protocol Labs has its own non-conventional things

Sometimes, because we have good reasons we haven't written down. Other times, :man_shrugging:. Never hesitate to ask.

_obligatory cat (mine and therefore the best in the world)_
![IMG_20190522_173855](https://user-images.githubusercontent.com/310393/61597544-cba56800-abc6-11e9-9bd7-bdfb3d0346d8.jpg)

**Comment by Stebalien** [comment]
@postables 

> How long do we have to give our final analysis? Ideally I'd like to test things for a week or so in dev but if that's too long thats understandable

Take your time. We'd like to get this out to users ASAP but we're _also_ trying out our new release process here and we want to get this right.

**Comment by postables** [comment]
@Stebalien understandable. It looks good right now, CI builds pass, no apparent new issues in dev, and so far no noticeable regressions. However if possible I'd like to hold off on a final judgement for a few more days in case anything crops up.

**Comment by obo20** [comment]
@Stebalien in response to your question: 

> Is this still happening (i.e., since July)?

We've been hitting this incredibly often and still do. It's got so bad that we started hosting our own copies of the binaries because dist.ipfs.io has gotten so bad.

**Comment by Stebalien** [comment]
Uploaded to GitHub.

**Comment by postables** [comment]
FWIW if you run into issues with sites like `dist.ipfs.io` you can quite easily load it up via a gateway like so: `https://foo.bar/ipns/dist.ipfs.io`

**Comment by koalalorenzo** [comment]
> FWIW if you run into issues with sites like `dist.ipfs.io` you can quite easily load it up via a gateway like so: `https://foo.bar/ipns/dist.ipfs.io`

Most of the time it doesn't work as the content itself is not easy to be discovered, unless magic with DHT or a direct connection. :( Hopefully a new version fixes that :P

**Comment by Stebalien** [comment]
This new version won't fix that, it's _just_ a patch release. We have some DHT patches that we believe will help once deployed to the entire network however, we're holding off till we can finish our test network so we can actually _test_ how this code will affect the network.

**Comment by ianopolous** [comment]
> @Stebalien in response to your question:
> 
> > Is this still happening (i.e., since July)?
> 
> We've been hitting this incredibly often and still do. It's got so bad that we started hosting our own copies of the binaries because dist.ipfs.io has gotten so bad (our ansible deployments were failing around 90% of the time due to timeouts)
> 
> The only reason we still encounter this problem is because we still have to initially pull new binary versions from dist.ipfs.io

For what it's worth, we hit the same issues, hence: https://github.com/peergos/ipfs-releases/

**Comment by Stebalien** [comment]
Early testers,

It's been a bit over a week. Any new issues with the release and/or can we move on to stage 3?

**Comment by b5** [comment]
tl;dr; LGTM

The Qri crew is completely tied up in non-IPFS stuff at the moment, leaving us little time to give proper feedback on this release cycle. I've taken a quick look at the changelog and everything is in keeping with what we've expected, so I'd rubber-stamp this as good-to-go.

We're very much looking forward to properly contributing to the early testing process on the next release. Please keep us in the loop!

**Comment by Stebalien** [comment]
@b5 SGTM. Thanks for the signoff.

**Comment by postables** [comment]
Totally forget to reply with my update, it looks good!

**Comment by koalalorenzo** [comment]
It looks good also for us on Siderus Orion client!

**Comment by Stebalien** [comment]
Stage 1... done!
Stage 2... done!
Stage 3...

Early testers:

We have entered stage 3 of our release process, the "soft" release. At this point, we consider this go-ipfs release to be production ready and we don't expect any more RCs at this point. Please deploy it on production infrastructure as you would a normal release. This stage allows us to rapidly fix any last-minute issues with the release without cutting an entirely new release.

- [ ] Infura (@MichaelMure)
- [ ] Textile (@sanderpick)
- [ ] Pinata (@obo20)
- [ ] RTrade (@postables)
- [ ] QRI (@b5)
- [ ] Siderus (@koalalorenzo)

When you're satisfied that 0.4.22-rc1 is at least as stable as 0.4.21, please sign off on this issue.

**Comment by sanderpick** [comment]
Same answer from me this time, @Stebalien. We're ahead of the release at the moment. Consider me a ✔️.

**Comment by obo20** [comment]
Same for us @Stebalien. Mark us as good to go.

**Comment by postables** [comment]
Looks good from my end, and even have a noticeable (albeit small) drop in CPU utilization :rocket: 
![lookin_good](https://user-images.githubusercontent.com/17089485/62172577-8af4cf80-b2e7-11e9-9fb9-69153d20e9d1.png)

**Comment by Stebalien** [comment]
That's probably:

> High bitswap CPU usage when connected to many (e.g., 10,000) peers. See ipfs/go-bitswap#154.

Good to know it helped.

**Comment by postables** [comment]
That would definitely cause it, good to know its working!

For what it's worth as well, there also appears to also be an improvement to memory usage as well. This is looking like a great release so far :rocket: 

![](https://gateway.temporal.cloud/ipfs/QmXAk2K3BUSHmkVCs7BYjsGcnNBSK47SknisTd643JNB6b)

**Comment by koalalorenzo** [comment]
Planned to deploy it tomorrow ( ~10:00 CET )

**Comment by Stebalien** [comment]
@koalalorenzo :fire: or :sunglasses:?

**Comment by Stebalien** [comment]
Stage 3 done.
Stage 4...

Building and releasing today (hopefully).

**Comment by hacdias** [comment]
@Stebalien status on this?

**Comment by Stebalien** [comment]
Built but we're waiting on some blog post stuff. We may release first if we can't get everything ready in time.

**Comment by rklaehn** [comment]
Just wanted to let you know that I very much agree with the decision put a hold on releasing new features until there is a process to ensure that the existing features work reliably...

**Comment by Retia-Adolf** [comment]
non-Windows binary in `go-ipfs_v0.4.22_windows-amd64.zip` :|

**Comment by Stebalien** [comment]
@Retia-Adolf thanks for the report. This should be fixed now and I apologize for flubbing it.

**Comment by andrewheadricke** [comment]
darwin amd64 build looks borked.

```
11:31:21.989 ERROR   cmd/ipfs: error from node construction:  could not build arguments for function "reflect".makeFuncStub (/usr/lib/go/src/reflect/asm_amd64.s:12): failed to build provider.Provider: could not build arguments for function "github.com/ipfs/go-ipfs/core/node".ProviderCtor (pkg/mod/github.com/ipfs/go-ipfs@v0.4.22/core/node/provider.go:24): failed to build *provider.Queue: function "github.com/ipfs/go-ipfs/core/node".ProviderQueue (pkg/mod/github.com/ipfs/go-ipfs@v0.4.22/core/node/provider.go:19) returned a non-nil error: strconv.ParseUint: parsing "1565442853283077000/b": value out of range daemon.go:337

Error: could not build arguments for function "reflect".makeFuncStub (/usr/lib/go/src/reflect/asm_amd64.s:12): failed to build provider.Provider: could not build arguments for function "github.com/ipfs/go-ipfs/core/node".ProviderCtor (pkg/mod/github.com/ipfs/go-ipfs@v0.4.22/core/node/provider.go:24): failed to build *provider.Queue: function "github.com/ipfs/go-ipfs/core/node".ProviderQueue (pkg/mod/github.com/ipfs/go-ipfs@v0.4.22/core/node/provider.go:19) returned a non-nil error: strconv.ParseUint: parsing "1565442853283077000/b": value out of range
```

**Comment by Stebalien** [comment]
@andreforsousa you've downgraded from master to 0.4.22. Master includes some new patches (and probably needs an explicit repo migration).

**Comment by andrewheadricke** [comment]
Thanks @Stebalien I deleted my `.ipfs` directory and now its working.
